pyccl/_repr.py

```diff
@@ -0,0 +1,383 @@
+import numpy as np
```

Comment:
I think all this looks ok, but I'd like us to have a library-level flag that determines whether to use these fancy reprs or just the standard "this is my id" repr. The default for this flag would be "Use the fancy reprs", but I'd like to easily switch to the simple option if needed (or if anyone determines that the performance hit actually matters for them).

Reply:
That's a good idea, I infused the `CCLObject` with an object similar to `_object_lock`, which I call `_fancy_repr`. It can be enabled/disabled using e.g. `ccl.CCLObject._fancy_repr.disable()` to fall back to Python's default memory-address-id.

Reply:
This is inherited, but we ideally want users who want to disable it (i.e. they know what they are doing) to only access it through `CCLObject` and *not* through any of its subclasses, or their instances. To prevent this from happening, I replace the `_fancy_repr` object in all subclasses of `CCLObject` with a descriptor that exits and throws a useful message of how to properly enable/disable the framework.

Reply:
Would it be so bad if users disabled it through the subclasses? I imagine in that case it's only disabled for that subclass (and its subclasses) right? And that might have its advantages (e.g. only disabling some of the expensive ones).

Reply:
Yes, it would be bad. There is only one instance of `FancyRepr` in the entire library. The classes that have been fancy-repred are stored in `FancyRepr._classes` so if someone were to do `ccl.Cosmology._fancy_repr.disable()` or even `cosmo._fancy_repr.disable()` they would in fact disable everything.

I would leave it if there were multiple instances of `FancyRepr` like they are with `ObjectLock` but I think this is an overkill for what we are trying to do.

Reply:
What I mean is that
```
id(ccl.Cosmology._fancy_repr) == id(cosmo._fancy_repr) == id(ccl.CCLObject._fancy_repr) == id(ccl.HaloProfileGaussian()._fancy_repr)
```

Reply:
Right, ok. Then, can you add a test that shows this failing?

Reply:
Ofc, I am writing all the tests at the moment.

Reply:
Done.